### ISSUE TITLE
Rounding errors in `Normed` to `Float` conversions

### ISSUE THREAD

**Issue by kimikage** [JuliaMath/FixedPointNumbers.jl#129]
While I was reviewing the PR #123, I found a minor bug.
```julia
julia> float(3N3f5) # this is ok
3.0f0

julia> float(3N4f12)
3.0000002f0

julia> float(3N8f8)
3.0000002f0

julia> float(3N10f6)
3.0000002f0

julia> float(3N12f4)
3.0000002f0

julia> float(3N13f3)
3.0000002f0

julia> float(3N2f6)
3.0000002f0

julia> float(3N4f4)
3.0000002f0

julia> float(3N5f3)
3.0000002f0
```
I think this is a problem with the division optimization.
https://github.com/JuliaMath/FixedPointNumbers.jl/blob/da3931838dd3a5f03e8ed00542a306d685002e33/src/normed.jl#L75-L78

The rounding errors can occur everywhere, but especially the errors at integer values are critical.
```julia
julia> isinteger(float(3N5f3))
false
```

**Comment by kimikage** [comment]
I was underestimating this issue. It is not so easy to fix this without performance regression.

WIP:
```julia
function (::Type{T})(x::Normed) where {T <: AbstractFloat}
    # The following optimization for constant division may cause rounding errors.
    # y = reinterpret(x)*(one(rawtype(x))/convert(T, rawone(x)))
    # Therefore, we use a simple form here.
    y = reinterpret(x) / convert(T, rawone(x))
    convert(T, y)  # needed for types like Float16 which promote arithmetic to Float32
end
function (::Type{T})(x::Normed{Ti,f}) where {T <: Union{Float32, Float16}, Ti <: Union{UInt8, UInt16}, f}
    f == 1 && return convert(T, reinterpret(x))
    r = Int64(reinterpret(x))
    s = UInt64(1)<<(64 - 8 * sizeof(Ti))
    k = floor(Int64, s / rawone(Normed{Ti,f})) # since `rawone()>=3>2`, `r*k` is always less than 2^63.
    y = convert(Float32, r * k) * (1.0f0 / s)
    convert(T, y)
end
```
The length of the mantissa part of `Float32` is `23 (+1)` bits, so `Int64` for the intermediate variable may seem excessive. However, I think the default rounding mode (*round half to even*) may require some *bitwise hacks*, and as a result, some conditional branches may be generated in the native code.
```julia
julia> Float32(0x01010101) == 0x1.010100p24 # I would like to round half up (`0x1.010102p24`), though.
true
```

Fortunately, the x86_64 CPUs have `CVTSI2SS xmm, r64`, which convert one `Int64` value to one `Float32` value and is not so slow.
However, with the other CPUs, I cannot imagine what the performance will be. In particular, the methods will be inlined, so I think the performance will depend greatly on the SIMD optimization.

Of course, only for `Normed{UInt8}`, the lookup tables may be nice.

**Comment by kimikage** [comment]
By the way, does anyone know a good way to avoid the method overwriting? Should I define methods using macros?

**Comment by kimikage** [comment]
Although the conversions from integer `Normed` values to `Float64` are OK, in some cases, the rounding errors can be noticeable.
```julia
julia> Float64(7.4N4f4)
7.3999999999999995

julia> Float64(reinterpret(7.4N4f4) / 15)
7.4
```

The results of the following method agree with the ideal values, but I think the following is not so good.
```julia
function Base.Float64(x::Normed{Ti,f}) where {Ti <: Union{UInt8, UInt16}, f}
    f == 1 && return Float64(reinterpret(x))
    r = Int64(reinterpret(x))
    s1 = UInt64(1)<<((53 - 16 + f - 1)÷f * f)
    s2 = UInt64(1)<<(((53 - 16)÷f + 1) * f)
    k = Int64(s1 ÷ rawone(Normed{Ti,f}))
    f64 = Float64(r * k)
    f64 * (1.0 / s1) + f64 * (1.0 / s1 / s2)
end
```

**Comment by kimikage** [comment]
In the case of `Normed{UInt8,f}` -> `Float32` where `f < 8` (i.e. except in the case of `Normed{UInt16}`), the length of the intermediate variable can be reduced to 32 bits as follows:
```julia
function Base.Float32(x::Normed{UInt8,2})
    r = Int32(reinterpret(x))
    s = 0b0100_0000_0000_0000 
    k = Int32(s ÷ rawone(Normed{UInt8,2}))
    Float32(r * k) * (1.0f0 / s + 1.0f0 / s / s)
end
```
However, on the x86_64 CPU with AVX/AVX2 extensions, the `Int32` version is slightly slower. In other words, although the modified version have an additional integer multiply, the `Int64` version is slightly faster than the current low-accuracy version, which is also use `Int32`.

**Comment by kimikage** [comment]
## Implementation status

|        from  \  to|Float16           |Float32           |Float64           |BigFloat          |
|:-------------------|:----------------:|:----------------:|:----------------:|:----------------:|
|Normed{UInt8}      |:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:|:grey_question:   |
|Normed{UInt16}     |:heavy_check_mark:|:heavy_check_mark:|:heavy_check_mark:|:grey_question:   |
|Normed{UInt32}     |                  |                  |                  |:grey_question:   |
|Normed{UInt64}     |                  |                  |                  |:grey_question:   |
|Normed{UInt128}    |                  |                  |                  |:grey_question:   |

**Comment by kimikage** [comment]
Do you have any ideas to simplify the following method, and to calculate it fast and perfectly, where `f in (17, 24, 25, 26, 29, 31)`?

```julia
function Base.Float64(x::Normed{UInt32,f}) where f
    f == 1 && return Float64(reinterpret(x))
    r = Int64(reinterpret(x))
    f ==  2 && return Float64(r * 0x040001) * 0x15555000015555p-72
    f ==  3 && return Float64(r * 0x108421) * 0x11b6db76924929p-75
    f ==  4 && return Float64(r * 0x010101) * 0x11000011000011p-72
    f ==  5 && return Float64(r * 0x108421) * 0x04000002000001p-75
    f ==  6 && return Float64(r * 0x09dfb1) * 0x1a56b8e38e6d91p-78
    f ==  7 && return Float64(r * 0x000899) * 0x0f01480001e029p-70
    f ==  8 && return Float64(r * 0x0a5a5b) * 0x18d300000018d3p-80
    f ==  9 && return Float64(r * 0x001001) * 0x080381c8e3f201p-72
    f == 10 && return Float64(r * 0x100001) * 0x04010000000401p-80
    f == 11 && return Float64(r * 0x000009) * 0x0e3aaae3955639p-66
    f == 12 && return Float64(r * 0x0a8055) * 0x186246e46e4cfdp-84
    f == 13 && return Float64(r * 0x002001) * 0x10000004000001p-78
    f == 14 && return Float64(r * 0x03400d) * 0x13b13b14ec4ec5p-84
    f == 15 && return Float64(r * 0x000259) * 0x06d0c5a4f3a5e9p-75
    f == 16 && return Float64(r * 0x011111) * 0x00f000ff00fff1p-80
    f == 18 && return Float64(r * 0x0b06d1) * 0x17377445dd1231p-90
    f == 19 && return Float64(r * 0x080001) * 0x00004000000001p-76
    f == 20 && return Float64(r * 0x000101) * 0x0ff010ef10ff01p-80
    f == 21 && return Float64(r * 0x004001) * 0x01fff8101fc001p-84
    f == 22 && return Float64(r * 0x002945) * 0x18d0000000018dp-88
    f == 23 && return Float64(r * 0x044819) * 0x07794a23729429p-92
    f == 27 && return Float64(r * 0x000a21) * 0x0006518c7df9e1p-81
    f == 28 && return Float64(r * 0x00000d) * 0x13b13b14ec4ec5p-84
    f == 30 && return Float64(r * 0x001041) * 0x00fc003f03ffc1p-90
    f == 32 && return Float64(r * 0x010101) * 0x00ff0000ffff01p-96

    f == 17 && return Float64(r) * 0x8000400020001p-68 # imperfect
    f == 24 && return Float64(r) * 0x1000001000001p-72 # imperfect
    f == 25 && return Float64(r) * 0x4000002000001p-75 # imperfect
    f == 26 && return Float64(r) * 0x10000004000001p-78 # imperfect
    f == 29 && return Float64(r) * 0x20000001p-58 # imperfect
    f == 31 && return Float64(r) * 0x80000001p-62 # imperfect
end
```

**Comment by kimikage** [comment]
# Benchmark
## Script
This script enables us to get a glimpse of problems of #125.
```julia
using BenchmarkTools
using FixedPointNumbers
struct Vec3{T <: Real}
    x::T; y::T; z::T
end
struct Vec4{T <: Real}
    x::T; y::T; z::T; w::T
end
Vec3{T}(v::Vec3{T}) where {T} = v
Vec3{T}(v::Vec3{U}) where {T, U} = Vec3{T}(v.x, v.y, v.z) 
Vec4{T}(v::Vec4{T}) where {T} = v
Vec4{T}(v::Vec4{U}) where {T, U} = Vec4{T}(v.x, v.y, v.z, v.w) 
Base.rand(::Type{Vec3{T}}) where {T} = Vec3{T}(rand(T), rand(T), rand(T))
Base.rand(::Type{Vec4{T}}) where {T} = Vec4{T}(rand(T), rand(T), rand(T), rand(T))
function Base.rand(::Type{T}, sz::Dims) where {T <: Union{Vec3, Vec4}}
    A = Array{T}(undef, sz)
    for i in eachindex(A); A[i] = rand(T); end
    return A
end
const N0f128 = Normed{UInt128,128}
const N125f3 = Normed{UInt128,3}
Ts = (N0f8, N5f3, N0f16, N13f3, N0f32, N8f24, N29f3, N0f64, N61f3, N0f128, N125f3)
vec1s = [rand(T,  64 * 64 * 4) for T in Ts];
mat3s = [rand(Vec3{T}, 64, 64) for T in Ts];
mat4s = [rand(Vec4{T}, 64, 64) for T in Ts];

for vec in vec1s
    println(eltype(vec), "-> Float32")
    @btime Float32.(view($vec,:))
end

for vec in vec1s
    println(eltype(vec), "-> Float64")
    @btime Float64.(view($vec,:))
end

for mat in mat3s
    println(eltype(mat), "-> Float32")
    @btime Vec3{Float32}.(view($mat,:,:))
end

for mat in mat3s
    println(eltype(mat), "-> Float64")
    @btime Vec3{Float64}.(view($mat,:,:))
end

for mat in mat4s
    println(eltype(mat), "-> Float32")
    @btime Vec4{Float32}.(view($mat,:,:))
end

for mat in mat4s
    println(eltype(mat), "-> Float64")
    @btime Vec4{Float64}.($mat)
end
```
## Julia v1.2.0 x86_64
### Vector
TBD.
### Matrix of Vec3
TBD.
### Matrix of Vec4
TBD.

**Comment by kimikage** [comment]
> Fortunately, all x86_64 CPUs have `CVTSI2SS xmm, r64`, which converts one `Int64` value to one `Float32` value and is not so slow. Most modern x86_64 CPUs have the AVX version `VCVTSI2SS`.

Although the conversion from `Int64` to `Float32` (with `VCVTSI2SS`, which is a scalar operation) itself is not so slow, it seems to degrade the performance. The compiler seems to abandon the SIMD vectorization when the `Int64` -> `Float32` conversion is used, and it can result in a long sequence of scalar operations. Unfortunately, the long sequences can interfere with inlining and make function calls, which have significant overheads.  `@inline` would be of no help since the SIMD vectorization is a matter outside of a single conversion method.

So, I'll try to avoid `Int64`->`Float32` whenever possible, even if it has the lower performance in a single conversion.

**Comment by timholy** [comment]
Wow, I've been focused on other tasks and missed all of this. Truly awesome analysis, @kimikage! 

> However, with the other CPUs, I cannot imagine what the performance will be.

I may own one of those older machines myself, but I think we should target the future more than the present.

> Do you have any ideas to simplify the following method, and to calculate it fast and perfectly, where f in (17, 24, 25, 26, 29, 31)

Not off the top of my head. Your lookup table idea is genius. I'm sure you know that the method will be statically optimized. What if we use a slow generic method for the troublesome cases and adopt your magic-number approach for the rest? At least for the common cases this should be a huge improvement.

This actually seems worth writing up as a paper. Have you considered that? I am a fine one to talk because I never get around to doing that for any of my Julia work, though I am trying to change that now.

**Comment by kimikage** [comment]
Thank you for your favorable feedback. My concern was (is) that this is out of balance between the benefit and the amount of codes. (I see this as a puzzle game, so I think this is worth the time *I* spent.)

> I'm sure you know that the method will be statically optimized.

Yes, I know that, and I just learned that Julia may give up the optimization due to just a little things. Although the compiling time and debugging of the C++ templates really disgust me, I'd like something like the `constexpr` keyword.

> What if we use a slow generic method for the troublesome cases and adopt your magic-number approach for the rest?

Yes, that is the plan. I'm going to finish this, but I will do my best for `N8f8`, `N6f10`, `N4f12`, `N2f14` and `N8f24`.

> This actually seems worth writing up as a paper. Have you considered that?

When I started this, I searched GitHub for some distinctive magic numbers, but I could not find a similar technique. (Well, the search engine of GitHub is somewhat poor, though.)
However, I think what I am actually doing is reinvent the wheel. I also investigated what codes the C++ compilers generate for the division by a constant. As the benchmark results above show, in short, the C++ compilers' answer is that the simple division with SIMD vectorization is the best! :jack_o_lantern:
My job is setting up the grave-post for the girl(?) who is not good at the SIMD optimization. I believe this will not be necessary in the near future.
Although I think my method has no academic value, but its background issues might be a little instructive to Julia programmers.

**Comment by timholy** [comment]
> Yes, I know that, and I just learned that Julia may give up the optimization due to just a little things. 

You could always write those methods as
```julia
i64(x) = Int64(reinterpret(x))
Base.Float64(x::Normed{UInt32,1}) = Float64(reinterpret(x))
Base.Float64(x::Normed{UInt32,2}) = Float64(i64(x) * 0x040001) * 0x15555000015555p-72
```
and then there's no risk they won't be optimized. Not really any longer, either.

**Comment by kimikage** [comment]
Well, the optimization for eliminating dead branches is quite excellent. I have no complaints or inconvenience about it. My complaint is more about inlining and compile-time calculation of constants.
For example, we know empirically that the `i64(x)` will be inlined. However, is it really obvious? The problem is that we cannot make sure it generally just by looking inside the method, i.e. it may depend on the caller or the caller's caller and so on. As you know and I mentioned above, `@inline` cannot control the inlining of its caller.
I came across an interesting issue about inlining and compile-time calculation, but I can't recall it. (I will show it when I can recall it.)

This is not a place to say a request, but I hope `exp2(::Integer)`, `ldexp(::Integer)`, etc. are calculated into constants at compile time.

**Comment by timholy** [comment]
Inlining is definitely a tricky & sensitive optimization. If you haven't found it, there's a bit of documentation about the internals: https://docs.julialang.org/en/latest/devdocs/inference/#The-inlining-algorithm-(inline_worthy)-1.

**Comment by kimikage** [comment]
I'm going to add a simple macro `@f32`.
```julia
macro f32(x::Float64) # for hexadecimal floating-point literals
    :(Float32($x))
end
```
I think it makes the macro more convenient to accept expressions as well as literals, but I'm not going to write a compiler.:smile:

> In the case of Normed{UInt16,f} -> Float32, where f < 8 || f == 16, the conversion method does not use the Int64 variable, and is suitable for the SIMD vectorization now.

For `N3f13`, `N2f14` and `N1f15`:
```julia
    i32 = Int32(reinterpret(x))
    f == 13 && return f32 * @f32(0x01p-13) + Float32(i32 * 0x07) * @f32(0x924db7p-52)
    f == 14 && return f32 * @f32(0x01p-14) + Float32(i32 * 0x15) * @f32(0xc30f3dp-56)
    f == 15 && return f32 * @f32(0x01p-15) + Float32(i32 * 0x49) * @f32(0xe071f9p-60)
```

**Comment by kimikage** [comment]
# Benchmark 2
## Julia v1.0.3 arm-linux-gnueabihf (Raspbian Buster on RPi2 v1.2)

```julia
julia> versioninfo()
Julia Version 1.0.3
Platform Info:
  OS: Linux (arm-linux-gnueabihf)
  CPU: ARMv7 Processor rev 4 (v7l)
  WORD_SIZE: 32
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, cortex-a53)
```
### Matrix of Vec4 (unit: μs)

|ARMv7|`Float32`<br><br>master|`Float32`<br>high<br>accuracy|`Float32`<br><br>simple div||`Float64`<br><br>master|`Float64`<br>high<br>accuracy|`Float64`<br><br>simple div|
|:-------|--------:|--------:|--------:|-|--------:|--------:|--------:|
|`N0f8`  | 249.843 | 100.312 |         | | 289.270 | 1.025e3 |         |
|`N5f3`  | 249.895 | 346.873 |         | | 290.260 | 1.007e3 |         |
|`N0f16` | 242.031 |  91.406 |         | | 263.333 | 1.031e3 |         |
|`N13f3` | 241.822 | 104.791 |         | | 263.125 | 1.025e3 |         |
|`N0f32` | 229.479 | 123.437 |         | | 246.719 | 107.500 |         |
|`N8f24` | 234.010 | 334.842 |         | | 246.875 | 410.155 |         |
|`N29f3` | 233.958 | 334.790 |         | | 245.937 | 107.447 |         |
|`N0f64` | 705.728 | 1.124e3 |         | | 998.072 | 1.123e3 |         |
|`N61f3` | 705.624 | 1.094e3 |         | | 998.957 | 1.092e3 |         |
|`N0f128`| 1.347e3 | 2.223e3 |         | | 1.584e3 | 2.158e3 |         |
|`N125f3`| 1.347e3 | 2.525e3 |         | | 1.584e3 | 2.440e3 |         |

**Comment by timholy** [comment]
Wow, those are all over the map. This is mostly due to the architecture or the Julia version?

**Comment by kimikage** [comment]
I think the inlining probrem is mostly due to the Julia version. However, I don't know which commit changed the behavior. (https://github.com/JuliaLang/julia/pull/24362 has been merged also into Julia v1.0.x branch.)
I'm going to try some workarounds (e.g. use of `@inline`, the pre-calculated literals, the LUTs with `const` keyword, the macros for generating bit patterns...).
BTW, I think the SIMD vectorization is so great that the inlining bonus should be added for the effect.

**Comment by kimikage** [comment]
### Note to Self: Workarounds for the inlining problem
- `@inline`
  - This is not so helpful since this has no effect to the caller.
  - This may be helpful for braked sub functions, because the optimization techniques (e.g. constant propagation, dead branch elimination) are applied when the sub functions are *inlined*.
    - Actually, the sub functions may not need the explicit `@inline`, though.
- Pre-calculated literals
  - This is very effective.
  - However, this introduces more magic numbers and verbose method definitions. 
- LUTs with `const` keyword
  - This is never good due to `inline_nonleaf_penalty`, even though the actual costs are very low.
- Macros for generating bit patterns
  - This may help remove *expensive* codes such as `/`.
    - `div_float` does not seem to be folded by Julia v1.0.x optimizer around the inlining stage, although it is finally folded.
  - This brings difficulties on readability and maintainability .

**Comment by kimikage** [comment]
> This is not a place to say a request, but I hope exp2(::Integer), ldexp(x,::Integer), etc. are calculated into constants at compile time.

Where Do We Come From? What Are We? Where Are We Going? :dizzy_face:
```julia
macro exp2(n)
     :(_exp2(Val($(esc(n)))))
end
@generated _exp2(::Val{N}) where {N} = (x = exp2(N); :($x))

# for Julia v1.0, which does not fold `div_float` before inlining
@generated inv_rawone(::T) where {T <: Normed} = (x = 1.0 / rawone(T); :($x))
```

**Comment by kimikage** [comment]
# Benchmark 3
## Julia v1.0.5 i686-w64-mingw32 (WoW64)

```julia
julia> versioninfo()
Julia Version 1.0.5
Commit 3af96bcefc (2019-09-09 19:06 UTC)
Platform Info:
  OS: Windows (i686-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
  WORD_SIZE: 32
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
```
### Vector (unit: μs)
|wow64|`Float32`<br><br>master|`Float32`<br>high<br>accuracy|`Float32`<br><br>simple div||`Float64`<br><br>master|`Float64`<br>high<br>accuracy|`Float64`<br><br>simple div|
|:-------|--------:|--------:|--------:|-|--------:|--------:|--------:|
|`N0f8`  |  60.600 |  60.600 |  60.700 | |  60.900 | *64.500*|  61.600 |
|`N5f3`  |  60.700 | *62.000*|  60.700 | |  60.900 | *65.600*|  61.600 |
|`N0f16` | 138.000 | 152.500 | 163.500 | | 136.100 | 150.400 | 172.300 |
|`N13f3` | 138.000 | 152.500 | 163.500 | | 136.100 | 150.600 | 172.400 |
|`N0f32` |  65.300 | *69.400*|  66.700 | |  64.800 |  64.900 |  65.500 |
|`N8f24` |  65.200 |*122.700*|  66.700 | |  64.800 |  66.500 |  71.600 |
|`N29f3` |  65.200 |*122.200*|  66.800 | |  64.800 |  66.300 |  69.900 |
|`N0f64` | 183.100 |  91.300 | 40.11e3 | | 161.100 |  92.100 | 41.25e3 |
|`N61f3` | 183.200 | 215.800 | 28.38e3 | | 161.100 | 197.600 | 28.25e3 |
|`N0f128`| 312.100 | 115.700 | 83.83e3 | | 328.600 | 117.900 | 83.61e3 |
|`N125f3`| 310.000 | 317.600 | 45.05e3 | | 328.800 | 314.800 | 45.17e3 |

`Normed{UInt32}`s are faster than native 64-bit version. `Normed{UInt64}`s and `Normed{UInt128}`s are slower. 

### Matrix of Vec4 (unit: μs)

|wow64|`Float32`<br><br>master|`Float32`<br>high<br>accuracy|`Float32`<br><br>simple div||`Float64`<br><br>master|`Float64`<br>high<br>accuracy|`Float64`<br><br>simple div|
|:-------|--------:|--------:|--------:|-|--------:|--------:|--------:|
|`N0f8`  |  18.300 |   3.263 |   3.350 | |  18.600 |   4.000 |   7.900 |
|`N5f3`  |  18.300 |   3.288 |   3.362 | |  18.600 |   3.671 |   7.900 |
|`N0f16` |  19.300 |   3.200 |   3.400 | |  19.600 |   3.957 |   7.933 |
|`N13f3` |  19.200 |   3.250 |   3.500 | |  19.600 |   3.733 |   7.900 |
|`N0f32` |  18.800 |   4.043 |   7.900 | |  18.500 |   5.050 |   7.933 |
|`N8f24` |  18.900 |   4.171 |   7.833 | |  19.000 |   7.633 |   7.967 |
|`N29f3` |  18.700 |   4.514 |   8.650 | |  22.500 |   5.750 |   8.033 |
|`N0f64` |  37.600 |  20.900 | 42.44e3 | |  41.000 |  20.000 | 42.44e3 |
|`N61f3` |  40.800 |  18.800 | 29.09e3 | |  39.800 |  18.100 | 28.05e3 |
|`N0f128`| 177.100 |  70.500 | 83.82e3 | | 220.000 |  69.900 | 83.91e3 |
|`N125f3`| 176.200 |  89.700 | 46.27e3 | | 218.900 |  85.700 | 45.70e3 |

`Normed{UInt64}`s and `Normed{UInt128}`s are slower than native 64-bit version. The others are faster or the same.

**Comment by kimikage** [comment]
BTW, I think there is a problem in interoperability with specialization using integer literal parameters.
```julia
julia> Normed{UInt32,Int64(1)} === Normed{UInt32,Int32(1)}
false

julia> Normed{UInt32,Int64(1)} == Normed{UInt32,Int32(1)}
false
```

**Comment by kimikage** [comment]
Well, you might wonder, "What about the `Fixed`?"
https://github.com/JuliaMath/FixedPointNumbers.jl/blob/ee5bd547bf73cf9ec6b976088c0c2565630f4ea5/src/fixed.jl#L63-L64
I could not believe my eyes. It far exceeds my understanding, so I'm not going to touch it.:see_no_evil:

**Comment by timholy** [comment]
Definitely not beyond your understanding, and indeed the Fixed cases are easier: they are scaled by 2^f rather than 2^f-1, so conversion operations are just bit-shifts. Here are the pieces:
- `TF(x.i>>f)`: this just converts the integer portion. `x.i>>f` truncates everything after the "virtual decimal point"
- `y = x.i&(one(widen1(T))<<f - 1)`: this zeros the bits corresponding to the integer portion, leaving only the fractional portion (the last `f` bits) represented as an integer
- `TF(y)/TF(one(widen1(T))<<f)`: this computes the fractional part as a floating-point number

**Comment by kimikage** [comment]
I apologize that I probably misled you. I mean..., I don't know the reason for not doing the following specialization.
```julia
Base.Float32(x::Fixed{T,f}) where {T <: Union{Int8, Int16}, f} = Float32(x.i) * (1.0f0 / (1<<f))
Base.Float64(x::Fixed{T,f}) where {T <: Union{Int8, Int16, Int32}, f} = Float64(x.i) * (1.0 / (1<<f))
```
Oops! :speak_no_evil:

**Comment by timholy** [comment]
I also wondered if it was more complicated than it needed to be. Assuming it works, seems worth changing. I've not really paid much attention (as you can tell) to `Fixed`.

**Comment by kimikage** [comment]
Fixed by PR #138.
The related issues found in this issue report and PR #138 will be considered separately.

**Comment by kimikage** [comment]
FYI, the problem with rounding error also occurs in `BigFloat` conversion. I am not going to fix this.
```julia
julia> 5.0N32f32 == 5.0
false

julia> Float64(5.0N32f32) == 5.0
true

julia> promote_type(N32f32, Float64)
BigFloat

julia> BigFloat(5.0N32f32)
5.000000000000000000000000000000000000000000000000000000000000000000000000000069
```
